### PR TITLE
Flattening

### PR DESCRIPTION
**Pull request by sharmishtaa**
Just pushing flattening code - adapted from Jun's confocal processing code. Haven't added a teest yet but wanted to push this trhough while I had some time.

### CODE REVIEW

File: acpreprocessing/flatten/flatten_utils.py [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+    "input_filename": "/Users/sharmishtaas/Documents/data/axonal/M6data/Section_13/ex1_2_13.tif",
+    "flatten_method": "top",
+    "output_filename": "test.tif",
+    "flip_back": False
+}
+
+def cleanImg(I):
+        mn = np.mean(I)
+        std = np.std(I)
+        I = np.where( (I1>(mn+ (3*std))), 0, I1)
```
[the remark this answers is not in the document]
Comment (kttakasaki) [reply acpreprocessing/flatten/flatten_utils.py:24]: Is cleanImg used anywhere? It looks like this is for removing outliers in noisy images - wouldn't expect this to be a problem for our detection.

**Review by kttakasaki — APPROVED** [review]

File: acpreprocessing/flatten/flatten_utils.py [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+                        ax.plot(range(len(curr_t)), curr_t, '-b', lw=0.5, alpha=0.1)
+
+                if curr_t[0] < surface_thr:
+                    curr_top = up_crossings(curr_t, surface_thr)
+                    if len(curr_top) != 0:
+                        top[yi, xi] = curr_top[0]
+
+                if curr_t[-1] < surface_thr:
+                    curr_bot = down_crossings(curr_t, surface_thr)
+                    curr_bot = curr_bot+30
```
[the remark this answers is not in the document]
Comment (sharmishtaa) [reply acpreprocessing/flatten/flatten_utils.py:159]: Yep, definitely makes sense for top as well, and as an argument.

File: acpreprocessing/flatten/flatten_utils.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (sharmishtaa) [reply acpreprocessing/flatten/flatten_utils.py]: Deleted .

File: acpreprocessing/flatten/flatten_utils.py [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+    if len(img.shape) != 3:
+        raise ValueError('input array should be 3d.')
+
+    z, y, x = img.shape
+
+    top = np.zeros((y, x), dtype=np.int)
+    bot = np.ones((y, x), dtype=np.int) * z
+
+    if is_plot:
+        f = plt.figure(figsize=(5, 5))
```
[the remark this answers is not in the document]
Comment (sharmishtaa) [reply acpreprocessing/flatten/flatten_utils.py:111]: added it in import line

File: acpreprocessing/flatten/flatten_utils.py [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+    find the index where the data up cross the threshold. return the indices of all up crossings (the onset data point
+    that is greater than threshold, 1d-array). The input data should be 1d array.
+    """
+    if len(data.shape) != 1:
+        raise ValueError('Input data should be 1-d array.')
+
+    pos = data > threshold
+    return (~pos[:-1] & pos[1:]).nonzero()[0] + 1
+
+def find_surface(img, surface_thr, is_plot=False):
```
[the remark this answers is not in the document]
Comment (sharmishtaa) [reply acpreprocessing/flatten/flatten_utils.py]: Not sure what this is referring to? both top and bottom are the same size as the image?

**Pull request merged**